controllers/config_listener.go

```diff
@@ -0,0 +1,159 @@
+package controllers
+
+import (
+	"fmt"
+
+	"github.com/infinispan/infinispan-operator/api/v2alpha1"
+	"github.com/infinispan/infinispan-operator/controllers/constants"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *infinispanRequest) ReconcileConfigListener() error {
+	if constants.ConfigListenerImageName == "" {
+		err := fmt.Errorf("'%s' has not been defined", constants.ConfigListenerEnvName)
+		r.log.Error(err, "unable to create ConfigListener deployment")
+		return nil
+	}
+	name := r.infinispan.GetConfigListenerName()
+	namespace := r.infinispan.Namespace
+	deployment := &appsv1.Deployment{}
+	objectMeta := metav1.ObjectMeta{
+		Name:      name,
+		Namespace: namespace,
+	}
+
+	err := r.Client.Get(r.ctx, types.NamespacedName{Namespace: namespace, Name: name}, deployment)
+
+	// The Deployment already exists, do nothing
+	if err == nil {
```

Comment:
During a GracefulShutdown upgrade the Deployment is deleted when `destroyResources` is called, however the old code didn't handle the HR upgrade scenario. I have updated so that if the ConfigListener image is different to the expected image, then the created resources are updated using the latest operator code.